### DISCUSSION TITLE
How to extract all Hausdorff distance data?

### DISCUSSION THREAD

**Discussion by Sukhe1993** [cnr-isti-vclab/PyMeshLab#34] [Q&A] [answered]
HI, I am using Ubuntu for the Meshlab.
I am now comparing two .stl surface model, I can obtain the HD in meshlab GUI and pymeshlab, but only the general data. which is like this:
`{'RMS': 1.5720561742782593,
 'diag_mesh_0': 129.02981567382812,
 'diag_mesh_1': 135.04889957361266,
 'max': 4.8711113929748535,
 'mean': 1.2516286373138428,
 'min': 0.00015201294445432723,
 'n_samples': 15774}`
I want to get all HD values, not only the mean, max, and min... Could you please indicate me how can I get the whole data in Meshlab GUI or Pymeshlab.

I also tried to open .ply file by using txt reader, but it is mojibake.
Thanks

**Comment by Sukhe1993** [comment]
Here i attached my code:

`import pymeshlab as ml
import samples_common
print('\n')
base_path = samples_common.samples_absolute_path()
out_path = samples_common.test_output_path()
ms = ml.MeshSet()
ms.load_new_mesh(base_path + "mtled_ventricle.stl")  
ms.load_new_mesh(base_path + "intra_ventricle.stl")  
res_dict = ms.apply_filter('hausdorff_distance', targetmesh=1, sampledmesh=0)`

output:

`{'RMS': 1.5720561742782593,
 'diag_mesh_0': 129.02981567382812,
 'diag_mesh_1': 135.04889957361266,
 'max': 4.8711113929748535,
 'mean': 1.2516286373138428,
 'min': 0.00015201294445432723,
 'n_samples': 15774}`

Thanks

**Comment by alemuntoni** [comment]
At the moment it is not possible to do perform the operation you need neither on meshlab nor on pymeshlab. This feature will be available in the future.

**Comment by cignoni** [comment]
Actually you can. :) 

You have to enable the "save samples" option in the filter and two point clouds will be created. They will be exactly of the same size and they contain the points sampled on the first surface and the corresponding closest point on the other surface, with quality storing the distance, so just save them and you have the full info.

**Comment by Sukhe1993** [comment]
> At the moment it is not possible to do perform the operation you need neither on meshlab nor on pymeshlab. This feature will be available in the future.

Thanks Alemuntoni.

**Comment by Sukhe1993** [comment]
> Actually you can. :)
> 
> You have to enable the "save samples" option in the filter and two point clouds will be created. They will be exactly of the same size and they contain the points sampled on the first surface and the corresponding closest point on the other surface, with quality storing the distance, so just save them and you have the full info.

Thanks Cignoni,  I have seen this in the _List of Filters_ and I tried `res_dict = ms.apply_filter('hausdorff_distance', targetmesh=1, sampledmesh=0, savesample = true)` but it does not works. Maybe I am wrong. Could you please upload the script? Thanks alot!
Yue

**Comment by alemuntoni** [comment, marked as answer]
Ok, I did not know that in the quality was stored the distance :)

In PyMeshLab, the MeshSet status it's actualy a "MeshLab session".
If in the MeshSet you have two meshes, the filter `hausdorff_distance` with `savesample = True` (be careful here: in python booleans are with the first capital letter) will create two other meshes with id 2 (Hausdorff Closest Points) and 3 (Hausdorff Sample Points).  You can then extract the quality which will contain the distance.
`vq = ms.mesh(2).vertex_quality_matrix()`

**Reply by mangobax** [reply]
I was looking for this, thanks. 
Just a heads up by looking into the documentation. The correct way of calling it would be `vq = ms.mesh(2).vertex_quality_array()`

Reference: https://pymeshlab.readthedocs.io/en/latest/classes/mesh.html

**Reply by alemuntoni** [reply]
yes, the name of the function changed since december!

**Reply by nurulistiqomahbudianti** [reply]
can you share the code,

**Reply by mangobax** [reply]
An example code would be:

Note: both meshes need to be aligned.

```
import pymeshlab  
ms = pymeshlab.MeshSet()
# Both meshes need to have the same orientation to be compared via Hausdorff
ms.load_new_mesh('copy_of_oriented_original.stl')
ms.load_new_mesh('copy_of_mesh_poisson.stl')
```
```
ms.apply_filter('hausdorff_distance', targetmesh=1, sampledmesh=0, savesample = True)
```
Using the above line results in something like this:
![firefox_FAEbScfEcf](https://user-images.githubusercontent.com/6492392/149441147-81474df8-0bbc-4eba-b77c-6f3b7bcbc09f.png)

```
vq = ms.mesh(2).vertex_quality_array()
```
vq has the histogram data from 2 samples compared via Hausdorff, so you may plot it if you need to:

```
import matplotlib.pyplot as plt
import numpy as np

# This is  the colormap I'd like to use.
cm = plt.cm.get_cmap('jet')

# Plot histogram.
n, bins, patches = plt.hist(vq, 256, density = True)
bin_centers = 0.5 * (bins[:-1] + bins[1:])

# scale values to interval [0,1]
col = bin_centers - min(bin_centers)
col /= max(col)

for c, p in zip(col, patches):
    plt.setp(p, 'facecolor', cm(c))

plt.xticks(np.linspace(float(format(min(vq), '.1f')),
               float(format(max(vq), '.1f')),
               num = 21,
               endpoint = True))
plt.tick_params(axis ='x', rotation = -90)

path = 'Some_path_in_your_drive'
output_file = path + "/HD_histogram.svg" # I save them in .svg format so later I can edit them on inkscape
plt.savefig(output_file,
        bbox_inches ="tight",
        transparent = True,
        facecolor ="w",
        edgecolor ='w',
        orientation ='landscape')    
    
plt.show()
```
![firefox_Rc5EFyCzKR](https://user-images.githubusercontent.com/6492392/149440875-695dc6fc-6e48-470e-ad45-9d54b146f0be.png)

**Comment by alemuntoni** [comment]
(BTW, I am moving this issue in the pymeshlab repo)

**Comment by CristianoPizzamiglio** [comment]
I apologize if I'm getting this wrong, but this discussion is a little confusing to me, i.e. @Sukhe1993 wrote:

> I want to get all HD values, not only the mean, max, and min

If I have properly understood the [Metro ](https://onlinelibrary.wiley.com/doi/abs/10.1111/1467-8659.00236) article by Dr. @cignoni, the values mapped in the user interface are the approximation errors, while the (asymmetric or directed) Hausdorff distance is the maximum given in the output result table. So you don't have many HD values for a single two-mesh comparison, but just one and only the asymmetric one.
Is this correct?
There is a little bit of confusion on [Stackoverflow ](https://stackoverflow.com/questions/65371819/meaning-of-hausdorff-distance-results-in-meshlab) as well.

Again, I apologize if I'm getting this all wrong. Thanks!
